### ISSUE TITLE
addItem calls a thread and that thread waits for the main thread

### ISSUE THREAD

**Issue by kelvinhammond** [openzim/libzim#692]
Hello,

I'm working on https://github.com/openzim/node-libzim/pull/72 and I noticed that after 10 items are inserted (before calling `finishZimCreation`) via `creator.addItem` on the main thread that a call is made via a thread to the main thread's creator.feed before that function returns. 

In nodejs there is an event loop which handles async operations. The problem with the main thread -> thread -> main thread while waiting on the main thread is that nodejs cannot respond because its busy in the main thread. Is there a way to turn this off or to have `addItem` return before the thread finishes running?

Output looks something like this from my debug.
What happens is
1. call to `creator.addItem(item)`
2. `ContentProvider.blob` is called, the calls `main:` is `MAIN_THREAD_ID == std::this_thread::get_id()`
3. a thread object is queued on the event loop but it does not return. I'll still debugging this part, will close this if its not relevant.
```
Starting
Created Zim, writing items...
ContentProvider::feed(), main: 1
ContentProvider::feed(), main: 1
0
ContentProvider::feed(), main: 1
ContentProvider::feed(), main: 1
1
ContentProvider::feed(), main: 1
ContentProvider::feed(), main: 1
2
ContentProvider::feed(), main: 1
ContentProvider::feed(), main: 1
3
ContentProvider::feed(), main: 1
ContentProvider::feed(), main: 1
4
ContentProvider::feed(), main: 1
ContentProvider::feed(), main: 1
5
ContentProvider::feed(), main: 1
ContentProvider::feed(), main: 1
6
ContentProvider::feed(), main: 1
ContentProvider::feed(), main: 1
7
ContentProvider::feed(), main: 1
ContentProvider::feed(), main: 1
8
ContentProvider::feed(), main: 0
From thread id: 140158619866688
Waiting on future
ContentProvider::feed(), main: 1
ContentProvider::feed(), main: 1
9
ContentProvider::feed(), main: 0
From thread id: 140158405965376
Waiting on future
ContentProvider::feed(), main: 0
From thread id: 140158628259392
Waiting on future
ContentProvider::feed(), main: 0
From thread id: 140158636652096
Waiting on future
```

**Comment by kelvinhammond** [comment]
I debugged a bit more

```
[New LWP 1517479]                                                                                                                                                                                                                      [9/416]
[Thread debugging using libthread_db enabled]                                                                                                                                                                                                 
Using host libthread_db library "/usr/lib/libthread_db.so.1".                                                                                                                                                                                 
0x00007fb5ecb21a55 in clock_nanosleep@GLIBC_2.2.5 () from /usr/lib/libc.so.6                                                                                                                                                                  
(gdb) bt                                                                                                                                                                                                                                      
#0  0x00007fb5ecb21a55 in clock_nanosleep@GLIBC_2.2.5 () from /usr/lib/libc.so.6                                                                                                                                                              
#1  0x00007fb5ecb26717 in nanosleep () from /usr/lib/libc.so.6                                                                                                                                                                                
#2  0x00007fb5e8213547 in zim::microsleep(int) () from /usr/lib/libzim.so.7                                                                                                                                                                   
#3  0x00007fb5e82202eb in ?? () from /usr/lib/libzim.so.7                                                                                                                                                                                     
#4  0x00007fb5e8216bea in zim::writer::CreatorData::closeCluster(bool) () from /usr/lib/libzim.so.7                                                                                                                                           
#5  0x00007fb5e8216ea6 in zim::writer::CreatorData::addItemData(zim::writer::Dirent*, std::unique_ptr<zim::writer::ContentProvider, std::default_delete<zim::writer::ContentProvider> >, bool) () from /usr/lib/libzim.so.7
#6  0x00007fb5e8218a81 in zim::writer::Creator::addItem(std::shared_ptr<zim::writer::Item>) () from /usr/lib/libzim.so.7
#7  0x00007fb5e83a93ca in Creator::addItem (this=0x55d8c5c28b60, info=...) at ../src/creator.h:132
.... I cut the rest here
```

If I look at the call to [`CreatorData::addItemData`](https://github.com/openzim/libzim/blob/master/src/writer/creator.cpp#L535) I can see the call to `closeCluster` which then calls `https://github.com/openzim/libzim/blob/48d16bd4e40eb4526c53082b8c4bceaf5b7401d4/src/writer/queue.h#L63` which sleeps the thread until the writer is done. How for nodejs this effectively blocks the event loop which does not offer it a chance to return the data that is needed for writing to disk. I've yet to figure out a solution to this. I could perhaps call `addItem` in a thread but I'm not quite sure that is enough.

**Comment by mgautierfr** [comment]
Which version of libzim are you using ? Please use nigthly to have https://github.com/openzim/libzim/pull/684

**Comment by kelson42** [comment]
@kelvinhammond Any feedback on this? We want to release 7.3.0 and this is essential libzim works fine for node-libzim.

**Comment by kelvinhammond** [comment]
I'll try the latest, I believe I'm on 7.2.0.

The treading issue here is during addItem libzim tries to wait until the queue has finished processing some items before it continues and blocks the main thread. I'll likely work around this somehow by calling addItem from a worker thread so it doesn't block the main thread's event loop. Which kind of defeats the purpose of addItem waiting in the queue and will cause a memory issue plus a huge amount of threads, so I need to see if there is a way to have the nodejs event loop process an iteration and continue somehow.

**Comment by mgautierfr** [comment]
I think master will not help here.
We have some kind of mutual lock. I suppose it is because js is not multithreaded and especially how js/you handle multithreading coming from libzim library.


<table>
<tr>
    <td> Main thread
    <td> Working thread
<tr>
	<td>
      <ul>
        <li> js call `addItem`
        <li> libzim (in `addItem`) check there is room in queue (10 places in total) and add task to queue
        <li> js give a chance to promise to run [??]
        <li> js loops and go to first point
        <li> When no more item to add, finish the creation in a  promise (and so give a chance to other promise to run for sure)
	<td>
      <ul>
        <li> libzim call `ContentProvider::feed()`
        <li> js create a promise (to be run in main thread)
        <li> sleep the current thread until the promise returns
        <li> When promise returns, return the value to libzim (and ends the task)
</table>

I'm not sure of the third point of the main thread.
If the promises created in workerthread are never run, the queue becomes full and the main thread wait for place to be able to add new task. But, as we need to exit `addItem` to be able to run the promise, this will never happens and we have a dead lock.
If you create a zim with less than 10 items, the queue is never full and the promises are run at the end (when we call finishZimCreation)

On python, this work as we release the gil when we call `addItem` so any call to `ContentProvider::feed()` in other thread are free to run
One solution would be to call `addItem` using a promise, this way it would not block call to `feed` and we are "sure" that the queue will be clear at a moment.

**Comment by kelvinhammond** [comment]
@mgautierfr I moved the `addItem` call to an `AsyncWorker`. This means I had to call back to `indexData` and `contentProvider` from the main thread too though since `addItem` was done on a thread. This works fine for now, I am currently debugging issues related to nodejs thread errors somewhere :/

**Comment by kelson42** [comment]
@kelvinhammond Good news, I guess that means we can close this ticket?

**Comment by kelson42** [comment]
@kelvinhammond Libzim 7.2.1 has been release. Closing this ticket.
